### ISSUE TITLE
SqlAGListener: When adding a second listener IP original IP gets mangled

### ISSUE THREAD

**Issue by jmos5156** [PowerShell/SqlServerDsc#1022]
**Details of the scenario you tried and the problem that is occurring:**
When building a multi-subnet failover cluster for a stretched AG, I add a new node from the new subnet to the WSFC fine. Using SqlAGlistener resource I add a new IP to the existing SQL listener during the DSC configurations of the new node. In doing so the original IP gets wiped and becomes 0.0.0.0/255.255.255.0 which mangles the cluster.

According to the logs, the SQL executes fine with the correct IP to be added listed.

**The DSC configuration that is using the resource (as detailed as possible):**

**Primary Node:**
```
                    if ($AvailabilityGroupRole -eq "Primary")
                    {
                        SqlAGListener "$($SQLNode)_$($SQLInstance)_$($AGGroup.Name)_listener"
                        {
                            Ensure = "Present"
                            ServerName = "r0-sqlag01pv,60001"
                            InstanceName = "r001_prod"
                            AvailabilityGroup ="r0-sqlag01pv"
                            Name = "r0-sqlag01pv"
                            IpAddress = @("10.30.2.254/255.255.255.0")
                            Port = 60001
                            PsDscRunAsCredential = $Credential_svc_deploy_ad
                        }
                    }
```
**Secondary Node :**
```
                    if ($AvailabilityGroupRole -eq "Secondary")
                    {
                        SqlAGListener "$($SQLNode)_$($SQLInstance)_$($AGGroup.Name)_listener"
                        {
                            Ensure = "Present"
                            ServerName = "r0-sqlag01pv,60001"
                            InstanceName = "r001_prod"
                            AvailabilityGroup ="r0-sqlag01pv"
                            Name = "r0-sqlag01pv"
                            IpAddress = @("10.40.2.254/255.255.255.0")
                            Port = 60001
                            PsDscRunAsCredential = $Credential_svc_deploy_ad
                        }
                    }
```

The actual code for SqlAGListener looks right an output to the log:

`
USE [master]
ALTER AVAILABILITY GROUP [r0-sqlag01pv]
MODIFY LISTENER N'r0-sqlag01pv'
(ADD IP (N'10.40.2.254', N'255.255.255.0')
);
`
Running the SQL script manually works fine.

Further to this running this PS code on the node hosting the primary replica also works:
```
$NodeName = "r0-sqlag01pv,60001"
$Name = "r0-sqlag01pv"
$AvailabilityGroup = "r0-sqlag01pv"
$InstanceName = "r001_prod"
$port = 60001

$sqlServerObject = Connect-SQL -SQLServer $NodeName -SQLInstanceName $InstanceName
$availabilityGroupObject = $sqlServerObject.AvailabilityGroups[$AvailabilityGroup]
$availabilityGroupListenerObject = $availabilityGroupObject.AvailabilityGroupListeners[$Name]

$IpAddress = "10.40.2.254/255.255.255.0"

$setListenerParams = @{
InputObject = $availabilityGroupListenerObject
StaticIp    = $IpAddress
}

Add-SqlAvailabilityGroupListenerStaticIp @setListenerParams
```

**Version of the Operating System, SQL Server and PowerShell the DSC Target Node is running:**
Windows 2016, SQL 2016 Ent, WMF 5.1

**What module (SqlServer or SQLPS) and which version of the module the DSC Target Node is running:**

**Version of the DSC module you're using, or 'dev' if you're using current dev branch:**
master

**Comment by jmos5156** [comment]
Running this code as an xScript resource in my DSC Config works confirming that something is not quite right with the DSC resource
```
$AGGroup.Name = "r0-sqlag01pv"
$AGGroup.ListenerName = "r0-sqlag01pv"
$AGGroup.TCPPort = 60001
$AGGroup.ListenerIpAddress = "10.40.2.254/255.255.255.0"
```

```
                        xScript "AddNewAGListenerIP"
                        {
                            GetScript = { }
                            TestScript = {return $false}
                            SetScript = {
                                $null = [System.Reflection.Assembly]::LoadWithPartialName('Microsoft.SqlServer.Smo')
                                $databaseEngineInstance = "$($using:AGGroup.Name),$($using:AGGroup.TCPPort)\$using:SQLInstance"
                                $sqlServerObject = New-Object Microsoft.SqlServer.Management.Smo.Server $databaseEngineInstance

                                $availabilityGroupObject = $sqlServerObject.AvailabilityGroups[$using:AGGroup.Name]

                                if ($availabilityGroupObject)
                                {
                                    $availabilityGroupListenerObject = $availabilityGroupObject.AvailabilityGroupListeners[$using:AGGroup.ListenerName]
                                    
                                    $presentIpAddress = $availabilityGroupListenerObject.AvailabilityGroupListenerIPAddresses
                                    
                                    $ipAddress = @()
                                    foreach ($currentIpAddress in $presentIpAddress)
                                    {
                                        $ipAddress += "$($currentIpAddress.IPAddress)/$($currentIpAddress.SubnetIPv4Mask)"
                                    }
                                    
                                    $missing = Compare-Object -ReferenceObject $using:AGGroup.ListenerIpAddress -DifferenceObject $ipAddress  | where-object {$_.SideIndicator -eq "<="}
                                    if ($missing)
                                    {
                                        Write-Verbose -Message 'Adding IP-address(es)'
                                        $newIpAddress += $missing.InputObject.Split(",").trim()
                                        
                                        $setListenerParams = @{
                                            InputObject = $availabilityGroupListenerObject
                                            StaticIp    = $newIpAddress
                                        }
                                        
                                        Add-SqlAvailabilityGroupListenerStaticIp @setListenerParams -ErrorAction Stop | Out-Null
                                    }	
                                }
                            }

                            PsDscRunAsCredential = $Credential_svc_deploy_ad
                        }
```

**Comment by johlju** [comment]
@jmos5156 Are you using SQLPS or SqlServer module (not SqlServerDsc)? If you are using SQLPS, then please try the SqlServer module. Either way, could you bug report this (referencing the correct module you are using) on the [PowerShell Gallery page](https://www.powershellgallery.com/packages/SqlServer) in the comment section. Also on the [SQL Server User Voice forum](https://feedback.azure.com/forums/908035-sql-server). Not sure which place is the best, so just submit on both, just in case. 😄

**Comment by jmos5156** [comment]
Good News. I fixed the bug and it was not to do with the resource. For a couple of days I was convinced that the issue was the DSC resource itself. I had introduced some code into the DSC Config last week as a test and forgot all about it. Naturally not writing anything to log, it all looked good except for the SQLListener DSC code which I had wrongly assumed was causing my issue.

So convinced was I with the repeating error especially when I used the code above confirming what I thought to be the root cause. It was only after I decided to strip back my template and add pieces back in step by step that I realised that my snippet of code from last week could have been the issue all along. 

Adding logging code to this confirmed my error.

Apologies for this :(
